### ISSUE TITLE
Remove possibility for shared state in PublicTasks and friends

### ISSUE THREAD

**Issue by agjohnson** [rtfd/readthedocs.org#3974]
Similar to the changes in #3946, we should break out the task steps to a
separate class to encapsulate state. Tests first will help determine if this is
prone to the same problem as the project.task Task subclasses.

**Comment by stsewd** [comment]
I only see that we use `PublicTask` here 

https://github.com/rtfd/readthedocs.org/blob/ba34164e28f56d0cb0ebc78cfb6ec801d0cae023/readthedocs/oauth/tasks.py#L24-L35

And we use that here 

https://github.com/rtfd/readthedocs.org/blob/ba34164e28f56d0cb0ebc78cfb6ec801d0cae023/readthedocs/restapi/views/task_views.py#L58-L59

Not sure if we depend on this code in other parts (like on the .com site),
but as right now, I'd say that we are sharing states when doing this

https://github.com/rtfd/readthedocs.org/blob/ba34164e28f56d0cb0ebc78cfb6ec801d0cae023/readthedocs/core/utils/tasks/public.py#L89-L89

Also, I couldn't find any other based class tasks.

**Comment by agjohnson** [comment]
Yeah, so this probably needs to be broken out into a stateful object instead of inside the task.

As far as i can remember, we only use the public task during remote repository syncing. I think we had planned on using it for build triggering, so we could update the status on the build list/output pages, but never got there.

**Comment by stsewd** [comment]
So, we access the request object inside the class, we can't do something like https://github.com/rtfd/readthedocs.org/pull/3946/ (or at least I think so :grimacing:). So, I thinking in using a bound task http://docs.celeryproject.org/en/latest/userguide/tasks.html#bound-tasks

**Comment by stsewd** [comment]
So, I was exploring this, I found that the same class is shared (same id), but the self.request object is different. I tested this creating two social applications and two users (one with GitHub and another with Gitlab) and using `rdb`. Maybe the `request` object isn't shared here? or I dit test this wrong?

Anyway, we can add this to https://github.com/rtfd/readthedocs.org/issues/3973 and be 100%  sure.
